treeherder/perf/models.py

```diff
@@ -107,32 +108,94 @@ class PerformanceDatumManager(models.Manager):
+        now = datetime.datetime.now()
+        elapsed_runtime = now - started_at
+
+        return max_overall_runtime < elapsed_runtime
+
+    def cycle_data(self, repository, cycle_interval, chunk_size, sleep_time,
+                   unsheriffed_repos_expire_days, logger,
+                   started_at, max_overall_runtime):
         """Delete data older than cycle_interval, splitting the target data
 into chunks of chunk_size size."""
-
-        max_timestamp = datetime.datetime.now() - cycle_interval
-
-        # seperate datums into chunks
-        while True:
-            perf_datums_to_cycle = list(self.filter(
-                repository=repository,
-                push_timestamp__lt=max_timestamp).values_list('id', flat=True)[:chunk_size])
-            if not perf_datums_to_cycle:
-                # we're done!
-                break
-            self.filter(id__in=perf_datums_to_cycle).delete()
-            if sleep_time:
-                # Allow some time for other queries to get through
-                time.sleep(sleep_time)
+        logger.warning('Cycling Perfherder data from {0} repository...'
+                       .format(repository.name))
+        max_timestamp = self._determine_max_timestamp(repository.name,
+                                                      cycle_interval,
+                                                      unsheriffed_repos_expire_days)
+
+        try:
+            self._delete_in_chunks(repository, max_timestamp, chunk_size, sleep_time, logger,
+                                   started_at, max_overall_runtime)
+        except NoDataCyclingAtAll as ex:
+            logger.warning('Exception: {}'.format(ex))
+            return
 
         # also remove any signatures which are (no longer) associated with
         # a job
+        logger.warning('Removing performance signatures with missing jobs...')
         for signature in PerformanceSignature.objects.filter(
                 repository=repository):
-            if not self.filter(signature=signature).exists():
+            if not self.filter(
+                    repository=repository,  # redundant, but leverages (repository, signature) compound index
+                    signature=signature).exists():
                 signature.delete()
 
+    def _delete_in_chunks(self, repository, max_timestamp, chunk_size, sleep_time, logger,
+                          started_at, max_overall_runtime):
+        from django.db import connection
+        any_succesful_attempt = False
+
+        with connection.cursor() as cursor:
+            while True:
+                if self._should_quit(started_at, max_overall_runtime):
+                    logger.info('Max runtime exceeded. Quitting cycling for repository {}'
+                                .format(repository.name))
+                    break
+
+                try:
+                    cursor.execute('''
+                        DELETE FROM `performance_datum`
```

Comment:
@sarah-clements I've decided to go for a raw query, because Django's ORM isn't capable of building DELETE w/ LIMIT queries.

Using the same cycling algorithm like that from Treeherder cannot delete more than 4000 rows/minute, which is way too slow for us. We would need 60 days just to be sure we keep Perfherder data under 500 billion rows.
From what I noticed, it's because we cannot simply increase the chunk_size to more than a few hundreds, when doing this; somehow, the query begins to freeze, then times out.

Anyway, the raw query seems to behave better with chunk_sizes of 10,000, but I need to test this 1st on treeherder-prototype.

Reply:
> I was talking to @klahnakoski about this and he thinks a good approach would be to set it to delete 1000 rows at a time (or the upper threshold of say 4000). It may indeed take a while to delete the old data but this way it will continuously delete the data at a rate that shouldn't cause issues (like a lock)- say every minute or 5 minutes.

You mean schedule the script to run every 1 or 5 minutes & only delete 1000 rows each time, then close?
This sounds like a safe solution & we should have considerable fewer lock issues.
Still, Heroku Scheduler can run every 10 minutes/1 hour/ day. This means we'll delete 144,000 rows (for chunk size at 1000) or 576,000 rows (for chunk size at 4000).

Reply:
Maybe a tight loop: inside is a transaction that will delete up to 1000 rows.  This will 
* delete more rows in a single run, 
* it can be simply scheduled once per day
* it will exit when there is nothing left to delete
* if it crashes then it will at least have done some work
* the small transactions will allow all other work to proceed as usual